whois-api/src/main/java/net/ripe/db/whois/api/rdap/RdapRelationService.java

```diff
@@ -0,0 +1,200 @@
+            case BOTTOM -> searchBottomResources(ipTree, searchIp);
+        };
+    }
+
+    private List<IpEntry> searchBottomResources(final IpTree ipTree, final IpInterval searchIp){
+        final List<IpEntry> mostSpecificValues = ipTree.findMostSpecific(searchIp);
+        final Set<IpEntry> mostSpecificFillingOverlaps = Sets.newConcurrentHashSet();
+        mostSpecificValues.forEach(ipResource -> extractBottomMatches(ipTree, searchIp, ipResource, mostSpecificFillingOverlaps));
+        return mostSpecificFillingOverlaps.stream().toList();
+    }
+
+    private void extractBottomMatches(final IpTree ipTree, final IpInterval searchIp,
+                                      final IpEntry mostSpecificResource,
+                                      final Set<IpEntry> mostSpecificFillingOverlaps) {
+
+        mostSpecificFillingOverlaps.add(mostSpecificResource);
+
+        final IpInterval mostSpecificInterval = IpInterval.parse(mostSpecificResource.getKey().toString());
+        final List<IpEntry> parentList = ipTree.findFirstLessSpecific(mostSpecificInterval);
+
+        if (parentList.isEmpty()){
+            return;
+        }
+
+        final List<IpEntry> siblingsAndExact = findSiblingsAndExact(ipTree, parentList);
+
+        final IpInterval firstSibling = (IpInterval)siblingsAndExact.getFirst().getKey();
+        final IpInterval lastSibling = (IpInterval)siblingsAndExact.getLast().getKey();
+
+        final IpEntry parent = parentList.getFirst();
+        final IpInterval parentInterval = (IpInterval) parent.getKey();
+
+        if (!parentInterval.equals(searchIp) && // If the parent is already the search ip we stop
+                !childrenCoverParentRange(firstSibling, lastSibling, parentInterval)){
+            extractBottomMatches(ipTree, searchIp, parent, mostSpecificFillingOverlaps);
+        }
+    }
+
+    private static boolean childrenCoverParentRange(final IpInterval firstResource, final IpInterval lastResource, final IpInterval parent){
+        return firstResource.beginAsInetAddress().equals(parent.beginAsInetAddress()) &&
+                lastResource.endAsInetAddress().equals(parent.endAsInetAddress());
+    }
+
+    private static List<IpEntry> findSiblingsAndExact(final IpTree ipTree, final List<IpEntry> parent) {
+        return ipTree.findFirstMoreSpecific(IpInterval.parse(parent.getFirst().getKey().toString()));
+    }
+
+    private IpEntry searchUpResource(final IpTree ipTree, final IpInterval searchIp){
+        final List<IpEntry> parentList = ipTree.findFirstLessSpecific(searchIp);
+        if (parentList.isEmpty() || !existAndNoAdministrative(searchIp, parentList.getFirst())){
+            throw new RdapException("404 Not Found", "No up level object has been found for " + searchIp.toString(), HttpStatus.NOT_FOUND_404);
+        }
+        return parentList.getFirst();
+    }
+
+    private IpEntry searchTopLevelResource(final IpTree ipTree, final IpInterval searchIp){
```

Comment:
I think if findAllLessSpecific getLast and its firstLessSpecific is already non administrative we can retun directly, instead of traversing whole tree because it can be really huge and its higher probability that you will  find within 2 levels